src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs

```diff
@@ -844,16 +844,95 @@ public override BoundNode VisitIsPatternExpression(BoundIsPatternExpression node
+            if (!VisitPossibleConditionalAccess(node.Expression, out var conditionalStateWhenNotNull)
+                || !trySetStateWhenNotNull(node.Pattern, conditionalStateWhenNotNull))
+            {
+                var expressionState = ResultType;
+                var state = PossiblyConditionalState.Create(this);
+                var labelStateMap = LearnFromDecisionDag(node.Syntax, node.DecisionDag, node.Expression, expressionState, ref state);
+                var trueState = labelStateMap.TryGetValue(node.IsNegated ? node.WhenFalseLabel : node.WhenTrueLabel, out var s1) ? s1.state : UnreachableState();
+                var falseState = labelStateMap.TryGetValue(node.IsNegated ? node.WhenTrueLabel : node.WhenFalseLabel, out var s2) ? s2.state : UnreachableState();
+                labelStateMap.Free();
+                SetConditionalState(trueState, falseState);
+            }
             SetNotNullResult(node);
             return null;
+
+            bool trySetStateWhenNotNull(BoundPattern pattern, PossiblyConditionalState conditionalStateWhenNotNull)
+            {
+                Debug.Assert(!IsConditionalState);
+
+                if (conditionalStateWhenNotNull.IsConditionalState)
+                {
+                    switch (getMatchingBoolValues(pattern))
+                    {
+                        case (hasTrue: true, hasFalse: false, hasNull: false):
+                            SetConditionalState(conditionalStateWhenNotNull.StateWhenTrue, State);
+                            return true;
+                        case (hasTrue: false, hasFalse: true, hasNull: false):
+                            SetConditionalState(conditionalStateWhenNotNull.StateWhenFalse, State);
+                            return true;
+                        case (hasTrue: false, hasFalse: true, hasNull: true):
+                            SetConditionalState(State, conditionalStateWhenNotNull.StateWhenTrue);
+                            return true;
+                        case (hasTrue: true, hasFalse: false, hasNull: true):
+                            SetConditionalState(State, conditionalStateWhenNotNull.StateWhenFalse);
+                            return true;
+                    }
+                }
+
+                switch (IsTopLevelNonNullTest(node.Pattern))
+                {
+                    case true:
+                        SetConditionalState(CloneAndUnsplit(ref conditionalStateWhenNotNull), State);
+                        return true;
+                    case false:
+                        SetConditionalState(State, CloneAndUnsplit(ref conditionalStateWhenNotNull));
+                        return true;
+                }
+
+                return false;
+            }
+
+            // Returns a struct with flags set to 'true' for each nullable boolean value that is accepted by the pattern.
+            static (bool hasTrue, bool hasFalse, bool hasNull) getMatchingBoolValues(BoundPattern pattern)
```

Comment:
Consider creating an enum and returning one of 5 values: MatchesTrue, MatchesFalse, MatchesNull, MatchesNotNull, Other.
It think that will be clearer, because we're not making use of the extra degrees of freedom of having 3 boolean flags.

Reply:
Yeah, to handle all the cases, I think it essentially needs to be a set.